### ISSUE TITLE
Unable to find the class named [java.sql.Date]

### ISSUE THREAD

**Issue by laeubi** [eclipse-ee4j/eclipselink#992]
I'm currently migrating an application form java 1.6 > java 11 but getting the following exception (can load java.sql.Date without a problem from a test-bundle), I'm using the latest 2020-12 Equinox release.

```
Caused by: Exception [EclipseLink-7156] (Eclipse Persistence Services - 2.7.7.v20200504-69f2c2b80d): org.eclipse.persistence.exceptions.ValidationException
Exception Description: Unable to find the class named [java.sql.Date]. Ensure the class name/path is correct and available to the classloader.
Internal Exception: java.lang.ClassNotFoundException: java/sql/Date
	at org.eclipse.persistence.exceptions.ValidationException.unableToLoadClass(ValidationException.java:2009)
	at org.eclipse.persistence.internal.jpa.metadata.MetadataHelper.getClassForName(MetadataHelper.java:115)
	at org.eclipse.persistence.internal.jpa.metadata.ORMetadata.getJavaClass(ORMetadata.java:334)
	at org.eclipse.persistence.internal.jpa.metadata.queries.StoredProcedureParameterMetadata.processArgument(StoredProcedureParameterMetadata.java:377)
	at org.eclipse.persistence.internal.jpa.metadata.queries.NamedStoredProcedureQueryMetadata.process(NamedStoredProcedureQueryMetadata.java:270)
	at org.eclipse.persistence.internal.jpa.metadata.MetadataProject.processQueries(MetadataProject.java:1716)
	at org.eclipse.persistence.internal.jpa.metadata.MetadataProcessor.addNamedQueries(MetadataProcessor.java:156)
	at org.eclipse.persistence.internal.jpa.EntityManagerSetupImpl.predeploy(EntityManagerSetupImpl.java:2036)
	... 22 more
Caused by: java.lang.ClassNotFoundException: java/sql/Date
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:398)
	at org.eclipse.persistence.internal.security.PrivilegedAccessHelper.getClassForName(PrivilegedAccessHelper.java:167)
	at org.eclipse.persistence.internal.jpa.metadata.MetadataHelper.getClassForName(MetadataHelper.java:101)
	... 28 more
```

**Comment by lukasj** [comment]
Any steps to reproduce the problem? Is java.sql explicitely required if applicable?

**Comment by laeubi** [comment]
What do you mean by 

> Is java.sql explicitely required if applicable?

The problem seems to be triggered by the following line

`@NamedStoredProcedureQuery(name = "NativeCheckQuery", procedureName = "FindCyclesOutbreak", parameters = { @StoredProcedureParameter(name = "AtDate", type = Date.class, queryParameter = "") })`

Date is java.sql.Data here, maybe a another type is possible to use? Its still a bit strange for a SQL-Persitence Framework that it can't use java.sql.* :-)

Also the code uses Timestamp in some places but I haven't yet got an explicit error about that but that might be becuase it fails already before?

**Comment by rfelcman** [comment]
It seems, that Java module `java.sql` is not available at module path. This is what @lukasj mentioned by 

> Is java.sql explicitely required if applicable?
Please check Your environment.
https://docs.oracle.com/en/java/javase/11/docs/api/java.sql/module-summary.html

**Comment by laeubi** [comment]
How do I check this? As mentioned above I can use java.sql classes in the same application from other places without a problem.

**Comment by rfelcman** [comment]
> How do I check this?

It's a question! Hmm from outside Java process I don't know. jConsole or JVisualVM doesn't provide functionality like this.
But let me rephrase my origin question.
Which environment is used there? It seems, that it's environment where every application runs in separate context with own classloader e.g. JEE applications in JEE servers.

**Comment by laeubi** [comment]
Ah okay I understand. I upgraded the application from an older/ancient Equinox/EclipseLink Version running Java 8, 
this now runs inside an OSGi (latest Equinox) container with latest 2.x EclipseLink on a full java 11 JDK without any special JVM options enabled.

**Comment by rfelcman** [comment]
Sorry but in this case please check Your Equinox settings. In our testing we e.g. Travis-CI we don't have issue with Java 11 and `java.sql.Date` class.

**Comment by laeubi** [comment]
I don't have issues either with Equinox +java.sql.Date class ;-)
But with this annotation EclipseLink fails to start (it can load/use java.sql in other places in the code as mentioned above!)

`@NamedStoredProcedureQuery(name = "NativeCheckQuery", procedureName = "FindCyclesOutbreak", parameters = { @StoredProcedureParameter(name = "AtDate", type = Date.class, queryParameter = "") })`

as soon as I comment this out everything is back to normal, so maybe I'm hitting a case here that is currently not covered by any tests?

**Comment by rfelcman** [comment]
I checked Equinox little bit and it seems, that there is something wrong in this OSGi container configuration . From message

`
...
Caused by: java.lang.ClassNotFoundException: java/sql/Date
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:398)
	at org.eclipse.persistence.internal.security.PrivilegedAccessHelper.getClassForName(PrivilegedAccessHelper.java:167)
...
`
it looks like, that `java.sql.Date` is not provided to application by Equinox.
I am not expert for the Equinox but is there any log?
Another diagnostic hint should be execute java process (with EclipseLink application) with `java -verbose:class ...` to get info about loaded classes and source jar files.
Next idea should be about some issue/collision with Java Module System. Possible test is to try JDK 8 (modules started with JDK 9), or add to java command line `.... --add-modules java.sql`it's not usually needed, but I don't know Your environment.

We are using `java.sql.Date` in EclipseLink tests (JDK 8, 11, 14, 15) very often and there are no any issues with this class loading.

**Comment by rfelcman** [comment]
Sorry, but for more than 1 month I didn't get response for latest questions. I'm closing this bug now, but feel free to reopen this if additional details are available.

**Comment by laeubi** [comment]
@rfelcman users without write-access can't reopen issues. There is simply no further information as already provided as mentioned above everything works fine except when using Date in an annotation like this:

@StoredProcedureParameter(name = "AtDate", type = Date.class, queryParameter = "") })

in other places java.sql.Date can be used without a problem, so I can only guess that this case is simply not covered by the existing test.

**Comment by rfelcman** [comment]
OK but as I played little bit with Equinox (standalone) it produce some log files and there should be some startup parameters. As I said I'm not Equinox specialist and without simulation I can't say what is wrong there. I'll reopen this bug now, but please provide there some Equinox details.

**Comment by rfelcman** [comment]
Maybe one difference is if  `java.sql.Date` in "normal" java code direct classloading is used, but in case of part of annotation like:

`@StoredProcedureParameter(name = "AtDate", type = Date.class, queryParameter = "") })`

is "secure" access used via
https://github.com/eclipse-ee4j/eclipselink/blob/2.7/jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/metadata/MetadataHelper.java#L101
https://github.com/eclipse-ee4j/eclipselink/blob/d751956f6453791f41fc6fe380231a108b11daa8/foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/internal/security/PrivilegedAccessHelper.java
In this case it should be related with security settings in Java or Equinox.

**Comment by laeubi** [comment]
@rfelcman thanks for the hint, security is disabled, I also see this in the debugger.
The classloader is org.eclipse.gemini.jpa.classloader.CompositeClassLoader wich looks fine so far with jdk.internal.loader.ClassLoaders$AppClassLoader as its parent. Could it be a difference because I'm using org.eclipse.persistence.annotations ?

**Comment by laeubi** [comment]
It seems the root cause is
https://bugs.eclipse.org/bugs/show_bug.cgi?id=572243
https://bugs.eclipse.org/bugs/show_bug.cgi?id=572244

so as long as the issue in Gemini is present I would assume that EclipseLink is doing everything fine, thanks for the patience and help!
